### DISCUSSION TITLE
Initramfs unpacking failed: invalid magic at start of compressed

### DISCUSSION THREAD

**Discussion by thesefer** [ipxe/ipxe#1329] [General]
Hi there

I have seen a discussion https://github.com/ipxe/ipxe/discussions/1244 and the issue https://github.com/ipxe/ipxe/issues/1312 which got closed. I am to open this topic again.

I am coming from 8f5fc161436a020ba65d07f91f62d34f4c22db61 with https://github.com/ipxe/ipxe/pull/612 implemented. After seeing https://github.com/ipxe/ipxe/commit/b1c13cc43ece3008f7012cf736fc943d5bb89131 I decided to rebuild without any manual patch.

![ipxe_after_dl_error](https://github.com/user-attachments/assets/8b030e8d-e2d8-4cc4-8b74-820e09babba3)

Unfortunately I am not able to install linux anymore (which works with my previous build+patch above) and with git bisect I identified the following change as the culprit:

6a004be0cceab5d669eedb5a2e6ee2feab31d5bd is the first bad commit

Please let me know if you need more information from me.

**Comment by NiKiZe** [comment]
Thanks for looking for existing issues before creating new ones. But as described in the linked discussions/issues, this is working as intended unless it kan be proven otherwise.

I suggest that you provide the contents of the script content you are using to boot the kernel.
As well as output from `imgstat` before `boot` to see what is loaded at that stage.
And finally the version of the Linux kernel that you are booting.

**Comment by thesefer** [comment]
Thanks for your feedback. From my point of view I went through the loops and invested my time to find the commit resulting in the failure of the deployment. For me I did the first step to proof that this change in iPXE broke something and not the kernel. But I am willing to collect more information If you are so kind to guide me through it.

Everything posted is unchanged from the default setup running for about 2 years now.
DHCP on MS Server 2019

Scope Policies
```
iPXE Configuration -> Boot\iPXE\iPXE.conf
UEFI x64 -> Boot\iPXE\iPXE64.efi
UEFI x86 -> Boot\iPXE\iPXE32.efi
UEFI x86 & x64 -> Boot\iPXE\iPXE64.efi
```

iPXE.conf
```
#!ipxe
chain --autofree Boot\iPXE\boot.ipxe.cfg
```

boot.ipxe.cfg
<details>

```
#!ipxe

# Menu settings
set menu-timeout 20000
isset ${menu-default} || set menu-default bootlocal

console --x 1024 --y 768 --picture /Boot\iPXE\/Wallpaper.png --left 200 --right 202 --keep
colour --rgb 0x4682b4 0 
colour --rgb 0x808080 1
colour --rgb 0xffffff 2
colour --rgb 0x000000 3
colour --rgb 0xff0000 5 # Text
colour --rgb 0xc0c0c0 6 # Separator
colour --rgb 0xffffff 7 # Standard color
cpair --foreground 3 --background 1 2

#iPXE Shell
cpair --foreground 7 --background 4 0

# Figure out if client is 64-bit capable
cpuid --ext 29 && set arch x64 || set arch x86
cpuid --ext 29 && set archl amd64 || set archl i386

[...] Menu stuff

:bootlocal
iseq ${platform} efi && goto boot-efi || goto boot-bios
goto MENU

:boot-efi
sanboot --no-describe --drive 0 || goto MENU

:boot-bios
sanboot --no-describe --drive 0x80 || goto MENU

:kubuntu2204-desktop
kernel http://dhcp.internal/rs-kubuntu-22.04-release/hwe-vmlinuz initrd=hwe-initrd ip=dhcp url=http://dhcp.internal/rs-kubuntu-22.04-release/rs-linux-22.04-curtin-installer-release.iso quiet locale=en_US languagechooser/language-name=English localechooser/supported-locales=en_US.UTF-8 countrychooser/shortlist=DE autoinstall ds=nocloud-net;s=http://repo.internal/.preseed/curtin-jammy/config-unattended-mobile/ root=/dev/ram0 cloud-config-url=/dev/null ---
initrd http://dhcp.internal/rs-kubuntu-22.04-release/hwe-initrd
boot
goto start
```

</details>

imgstat
```
:iPXEDebug
kernel http://dhcp.internal/rs-kubuntu-22.04-release/hwe-vmlinuz initrd=hwe-initrd ip=dhcp url=http://dhcp.internal/rs-kubuntu-22.04-release/rs-linux-22.04-curtin-installer-release.iso quiet locale=en_US languagechooser/language-name=English localechooser/supported-locales=en_US.UTF-8 countrychooser/shortlist=DE autoinstall ds=nocloud-net;s=http://repo.internal/.preseed/curtin-jammy/config-unattended-mobile/ root=/dev/ram0 cloud-config-url=/dev/null ---
initrd http://dhcp.internal/rs-kubuntu-22.04-release/hwe-initrd
itemstat
goto shell
```

Result
```
hwe-vmlinuz... ok
hwe-initrd... ok
Wallpaper.png : 130327 bytes [PNG]
hwe-vmlinuz : 12186376 bytes [EFI] [SELECTED] "initrd=hwe-initrd ip=dhcp url=http://dhcp.internal/rs-kubuntu-22.04-release/rs-linux-22.04-curtin-installer-release.iso quiet locale=en_US languagechooser/language-name=English localechooser/supported-locales=en_US.UTF-8 countrychooser/shortlist=DE autoinstall ds=nocloud-net;s=http://repo.internal/.preseed/curtin-jammy/config-unattended-mobile/ root=/dev/ram0 cloud-config-url=/dev/null ---"
hwe-initrd : 121462526 bytes
```

`imgstat` shows the same result using https://github.com/ipxe/ipxe/commit/8f5fc161436a020ba65d07f91f62d34f4c22db61 and the first bad commit https://github.com/ipxe/ipxe/commit/6a004be0cceab5d669eedb5a2e6ee2feab31d5bd

The drop in shell after loading the iso gives the following kernel: 5.19.0-32-generic x86_64

**Reply by mcb30** [reply]
> ```
> hwe-vmlinuz... ok
> hwe-initrd... ok
> Wallpaper.png : 130327 bytes [PNG]
> hwe-vmlinuz : 12186376 bytes [EFI] [SELECTED] "initrd=hwe-initrd ip=dhcp url=http://dhcp.internal/rs-kubuntu-22.04-release/rs-linux-22.04-curtin-installer-release.iso quiet locale=en_US languagechooser/language-name=English localechooser/supported-locales=en_US.UTF-8 countrychooser/shortlist=DE autoinstall ds=nocloud-net;s=http://repo.internal/.preseed/curtin-jammy/config-unattended-mobile/ root=/dev/ram0 cloud-config-url=/dev/null ---"
> hwe-initrd : 121462526 bytes
> ```

The issue is the presence of `Wallpaper.png`.  Free this using `imgfree` before the `kernel` command, and all should be well.  Alternatively, use the default behaviour of the `console --picture ...` command, which will be to automatically free the image (unless you explicitly specify the `--keep` option).

**Reply by mcb30** [reply]
For more background: the commit https://github.com/ipxe/ipxe/commit/6a004be0cceab5d669eedb5a2e6ee2feab31d5bd that you identified changes the behaviour so that the initramfs provided under UEFI matches the initramfs that would have been provided under BIOS.  If you try using your iPXE boot script on a BIOS system, you should find that the same error occurs (either before or after commit https://github.com/ipxe/ipxe/commit/6a004be0cceab5d669eedb5a2e6ee2feab31d5bd).

**Reply by NiKiZe** [reply]
Yes good base research and narrow down where the change happened, awesome actually!

It would have been even better if the information that was requested in the linked discussions was provided as well.

In the end the change is expected and the issue you are seeing, as explained, is due to how your script is setup. The fix is to adjust the script as explained by mcb30.

**Reply by thesefer** [reply]
Thanks for your explanation! Looks like I missed some crucial information while doing my research, apologies! I did a quick test without any further changes by simply removing `--keep` for now. Now I have troubles to build it properly using named configurations, but this is part of another discussion I found.

I will close the discussion for now because the issue has been identified and resolved. I hope it helps others stumbling across this.

**Reply by WaterSlide** [reply]
imgfree before kernel helped solve my issue. Just wanted a fancy background image, I didn't think it would take all day to solve. Glad to see this within the same 24 hour period that I had the same issue!

**Reply by redat00** [reply]
Hi ! 

I think I'm going to need a little bit more information about what `imgfree` does and why it repaired a failing boot process..

We have two different OS that we tried to boot using iPXE : One is a classic Debian based, and the other one is a archlinux one. Both of them were encountering a decompression error on the initramfs.. And both of them did boot successfully once we added the `imgfree` command just before it. 

The strangest thing ? When the same network card is connected on 1G link instead of the 25G link we're trying to boot it on : It doesn't require the `imgfree` command...

It's clear we have a pattern here, but I just want to see what exactly does the `imgfree` command is supposed to help ? And just to be clear : We don't use any kind of background image or anything like this. Just plain iPXE script. Nothing more, nothing less.

Thanks,
Renaud.

**Reply by NiKiZe** [reply]
Use `imgstat` before your `imgfree` then do `imgstat` again, and finally use `imgstat` before `boot`

Do this on your 2 scenarios, make sure you note down the version of iPXE you are using, as well as if you are using pcbios or efi. (Are you sure it is only the link speed that changes, and no other part of iPXE binaries or scripts?)

If you need further explanation between your 1G and 25G case, please post the output from data collected above.

**Reply by redat00** [reply]
Hello ! 

I'm so relieved we found the difference and what was actually the problem. Our 2 setup were not exactly equal : 
- One was downloading a `autoexec.ipxe` file
- The other wasn't (as it didn't exist)

Removing the `autoexec.ipxe` file make the system work, without a `imgfree` command ! 

The only question I have is : Where does those get downloaded ? How much space is there where this is ? The file is only 17 bytes, and I don't think that it would matter in size that much that the rest couldn't be downloaded. Am I missing something ? 

Thanks for your help! 

Renaud.

**Reply by NiKiZe** [reply]
Good it resolved your issue.
Is it a valid cpio archive that the kernel can understand? (No it is a ipxe script.)
It's not about size it's about it being used as part of a generated cpio archive. This simply corrupts the data the kernel uses, (the scripts told ipxe to do so)

It would be clearer, and easier to explain if the output of `imgstat` was given.

**Comment by pjotrp** [comment]
Hey, thanks for ipxe. I spent half a day trying to fix the initrd.gz failure. This thread gave the pointers to fix it. I also had an autoexec.ipxe chainer.
